pkg/scd/uss_availability_handler.go

```diff
@@ -2,34 +2,108 @@ package scd
 	"github.com/interuss/dss/pkg/api/v1/scdpb"
 	dsserr "github.com/interuss/dss/pkg/errors"
+	dssmodels "github.com/interuss/dss/pkg/models"
+	scdmodels "github.com/interuss/dss/pkg/scd/models"
+	"github.com/interuss/dss/pkg/scd/repos"
 	"github.com/interuss/stacktrace"
+	"strings"
 )
 
+func GetDefaultAvailabilityResponse(id dssmodels.Manager) *scdpb.UssAvailabilityStatusResponse {
+	return &scdpb.UssAvailabilityStatusResponse{
+		Status: &scdpb.UssAvailabilityStatus{
+			Availability: "Unknown",
+			Uss:          id.String()},
+		Version: "",
+	}
+}
+
 func (a *Server) GetUssAvailability(ctx context.Context, request *scdpb.GetUssAvailabilityRequest) (*scdpb.UssAvailabilityStatusResponse, error) {
-	id := request.GetUssId()
+	id := dssmodels.ManagerFromString(request.GetUssId())
 	if id == "" {
 		return nil, stacktrace.NewErrorWithCode(dsserr.BadRequest, "UssId not provided")
 	}
-	response := &scdpb.UssAvailabilityStatusResponse{
-		Status: &scdpb.UssAvailabilityStatus{
-			Availability: "Unknown",
-			Uss:          id},
+
+	var response *scdpb.UssAvailabilityStatusResponse
+	action := func(ctx context.Context, r repos.Repository) (err error) {
+		// Get USS availability from Store
+		ussa, err := r.GetUssAvailability(ctx, id)
+		if err != nil && err != sql.ErrNoRows {
+			return stacktrace.Propagate(err, "Could not get USS availability from repo")
+		}
+		if ussa == nil {
+			// Return default availability status "Unknown"
+			response = GetDefaultAvailabilityResponse(id)
+			return nil
+		}
+		response = &scdpb.UssAvailabilityStatusResponse{
+			Status: &scdpb.UssAvailabilityStatus{
+				Availability: ussa.Availability.String(),
+				Uss:          id.String()},
+			Version: ussa.Version.String(),
+		}
+		return nil
+	}
+
+	err := a.Store.Transact(ctx, action)
+	if err != nil {
+		if strings.Contains(err.Error(), "does not exist") {
+			response = GetDefaultAvailabilityResponse(id)
+		} else {
+			return nil, err // No need to Propagate this error as this is not a useful stacktrace line
+		}
 	}
 	return response, nil
 }
 
 func (a *Server) SetUssAvailability(ctx context.Context, request *scdpb.SetUssAvailabilityRequest) (*scdpb.UssAvailabilityStatusResponse, error) {
-	id := request.GetUssId()
-	if id == "" {
-		return nil, stacktrace.NewErrorWithCode(dsserr.BadRequest, "UssId not provided")
+	return a.PutUssAvailability(ctx, request.GetUssId(), "", request.GetParams())
+}
+
+func (a *Server) PutUssAvailability(ctx context.Context, ussID string, version string, params *scdpb.SetUssAvailabilityStatusParameters) (*scdpb.UssAvailabilityStatusResponse, error) {
+	if ussID == "" {
+		return nil, stacktrace.NewErrorWithCode(dsserr.BadRequest, "ussID not provided.")
 	}
-	result := &scdpb.UssAvailabilityStatusResponse{
-		Status: &scdpb.UssAvailabilityStatus{
-			Availability: "Unknown",
-			Uss:          id},
+	// Retrieve USS availability status from request params
+	availability := params.GetAvailability()
+	if availability == "" {
```

Comment:
Let's also check against the allowed values for availability -- right now, it seems like a USS could set `uss1`'s availability to `Purple`.  A good way to do this would probably be to write `scdmodels.UssAvailabilityStateFromString(string)` and have it return `(UssAvailabilityState, Error)`, and populate that error if the string didn't match one of the allowed values.  Bonus points if we parse without case sensitivity, but return a properly-capitalized value :)  The check for `if availability == ""` would presumably also go into `scdmodels.UssAvailabilityStateFromString`.

Reply:
Done